### ISSUE TITLE
Redundant computation transformation

### ISSUE THREAD

**Issue by rupertford** [stfc/PSyclone#50]
By default PSyclone will generate loop bounds which correctly compute locally owned elements.

A potential optimisation would be to redundantly compute values in the halos. The potential advantage of this is that we might avoid (halo exchange) communication if we do so.

The Met Office have identified support for redundant computation as a priority.

This issue will discuss and then implement a redundant computation transformation for lfric.

**Comment by rupertford** [comment]
I think redundant computation in the halos can be controlled by modifying the loops upper bound using the mesh's get_last_halo_cell(n) function, where n is the depth of the halo i.e.
When iterating over cells I think redundant computation in the halos can be controlled by modifying the loops upper bound using the mesh's get_last_halo_cell(n) function, where n is the depth of the halo i.e.

DO cell=1,mesh%get_last_halo_cell(n)

where n>0 and n<= mesh%get_halo_depth()

If we would like to compute in all levels of the halo, but do not know what depth that is, I think we can do the following

DO cell=1,mesh%get_last_halo_cell(mesh%get_halo_depth())

When iterating over dofs I think redundant computation in the halos can be controlled by modifying the loops upper bound using the field proxy's get_last_dof_halo(n) function, where n is the depth of the halo i.e.

DO df=1,v_proxy%vspace%get_last_dof_halo(n)

Again, if we would like to compute in all levels of the halo, but do not know what depth that is, I think we can do the following ...

DO df=1,v_proxy%vspace%get_last_dof_halo(mesh%get_halo_depth())

I think it would be good if @mike-hobson would check the above.

**Comment by mike-hobson** [comment]
Hi Rupert,

Your information on looping over cells is correct.

Unfortunately, the information when iterating over dofs isn’t. The call to get_last_dof_halo() in the function space object presently only returns the last dof in deepest halo. It would be a straightforward task to overload this with a function that takes halo depth as an argument, if this is needed.

Performing redundant computation deep into the halos should be done with care. We only designed the infrastructure to support full redundant computation out to the first depth of halo. So, for example, no operation that requires a stencil bigger than a single cell will work in the deepest halos.

Hope that helps,
Mike.

**Comment by rupertford** [comment]
Thanks for the info @mike-hobson. My immediate reaction to your question about needing redundant computation at different levels of the halo for dofs is yes, I think we probably will want fine grain control at some point. All the internals are there in the infrastructure already (the underlying storage is an array of size halo depth!) so, assuming the dofs are in halo depth order (is that something that occurs naturally when ensuring cells are in halo depth order?) then, as you say, it should be simple to add, and is already coded for!

As for computing deep into the halos, I think we need to determine what the limitations are here so I can ensure we only allow what is supported. Could you please elaborate i.e. give me the rules. For example you might be saying that computation with no stencil can compute out to any halo depth and that computation with a stencil of depth one can compute out to the full depth-1 but that computation with a stencil of depth > 1 can only compute out to a halo of depth 1.

**Comment by rupertford** [comment]
All transfomations sub-class the Transformation class and must implement the apply method. I suggest that we call this transformation

``
DynamoRedundantComputationTrans
``

(very catchy!) and that it's apply interface takes the following form

``
schedule, memento apply(node, depth=n)
``

`schedule` and `memento` are standard return values, `node` should be of type `loop` and the `depth` argument is optional. If the `depth` argument is not provided then redundant computation will be performed to the full halo depth (taking into account stencil accesses). The value of `n` should be a positive integer.

If the loop is writing to a continuous function and iteration over cells then the value of `n` must be at least 2 (as it computes in the level 1 halo by default).

We should probably add a check to test whether `n` (plus stencil) is greater than the maximum halo depth. This could be a run-time check that aborts the code, or perhaps a warning and reduces the computation to a valid value. An alternative would be to provide PSyclone with the maxmum halo depth so we could check at generation time.

**Comment by mike-hobson** [comment]
Yes, all the infrastructure is in place to provide a  get_last_dof_halo(n) function – it’s just a case of adding the getter. I have opened LFRic ticket #1092 to add it.

I think the complication with redundant computation in the halo is exactly as you describe. You can only do and operation using a stencil of depth n (where a single cell stencil has a depth of 0) out to halo (full_depth-n).

**Comment by rupertford** [comment]
OK, I think I have enough info to get started.

**Comment by rupertford** [comment]
The way in which halo exchanges are currently coded does not allow us to generate correct halo depths in a number of cases when we have a field in more than one kernel depending on the same halo exchange. I've created a new issue for this (#62) as this issue is about supporting redundant computation. The reason the problem has come up here is that we have re-written the halo placement code which now only places the minimum number of halos required.

**Comment by rupertford** [comment]
I've created a discussion ticket on SRS (1152) to decide whether out-of-bounds checking when performing stencil accesses and/or redundant computation should be done by PSyclone, by the halo exchange call, or a combination of the two. I think I prefer the last option which would require additional checking code by PSyclone for cases where there are only writes to fields in a kernel. Typically there will not be any in standard kernels but there will be when using some builtin's. Unless I create another issue for this, I will have to wait until the issue is resolved before creating a pull request to master.

**Comment by rupertford** [comment]
I think the discussion ticket on SRS (1152) has got far enough to come to a conclusion. The conclusion is that halo_exchanges already check that halo accesses are not out-of-bounds. These checks ensure that a read to a field in a kernel will not be out-of-bounds. However, PSyclone will need to add code to check that redundant writes to fields will not be out-of-bounds. I've created a new issue #64 to implement such checks.

Related to this discussion, I also realised that PSyclone does not currently check the validity of stencil extent values passed from the algorithm layer. For example, a user could specify a negative halo and it would not raise an exception at run-time. I have therefore created a new issue #65 which will address this problem.

**Comment by rupertford** [comment]
I wrote an example which used redundant computation to remove the need for halo exchanges (or move the halo exchanges). This is the main reason for supporting redundant computation. However, the current implementation does not remove halo exchanges. When thinking about this problem it came back to the same issue as #62 and in order to make progress I think I need to address this problem.

Essentially we need a way to keep track of reads from halos for a field when they all depend on the same halo exchange (this is #62). We also need to keep track of writing to a halo so we know whether a particular halo exchange is needed or not.

The way I propose to do this is for a halo exchange to keep a list of all readers and for us to be able to add, remove, or modify readers from that list. We will also keep a single entry for the writer and will be allowed to modify this writer. It will then be up to the halo exchange to generate appropriate halo bounds from its list. Given this functionality we can then follow the logic below

```
for each potential read to a halo in a loop:
    if not previous dependence:
        add new halo exchange
    else:
        if previous dependence not a halo exchange:
            add new halo exchange
        else:
            if registered with halo exchange:
                update my information
            else:
                add my information
```

Notice that the main loop is over all halo reads (not unique which is currently the case). This solves the problem of having two or more halo reads of different sizes in the same loop. Note I think this would be a problem but have not yet checked - I presume there is no test for this case at the moment.

Notice we do not look at reads with no halos. In theory someone could change a loop so that something that had a halo access previously, no longer did so. e.g. go from halo(1) to ncells. However, there are no current transformations that will do this and I can't think of an example that would so we are safe at the moment.

We would also need to add the following logic for writers ...

```
for each write to halo:
    if there is a following dependence:
        if the dependence is a halo exchange:
            update my details
            if halo no longer required:
                remove halo exchange
```

The above logic skips the case where we have a write dependence but no halo exchange. Clearly we would need a halo exchange here but the logic works as we first add halo exchanges before any transformations are applied and when we add these we iterate over all loops, so all halo exchanges are added by the read logic.

**Comment by arporter** [comment]
All points have been addressed (apart from the docs but I've done that now). Tests all pass. pep8 and pylint are clean.
Merged #78 manually to master. In retrospect that might hamper github's attempts to fix the issue with it but I didn't expect it to automatically close the PR.
